Add block cache for low level table library

byronyi opened the pull request:
This is part of a patch series aiming to improve the performance of on-disk dataset.cache() (CacheDatasetV2).

Currently CacheDataset uses core/util/tensor_bundle to cache dataset elements on disks. It uses sorted string table (SST) to index dataset elements. Unlike checkpoints which do not have a great number of tensors, caching a large dataset may incur a greater number of tensors as well as index blocks.

If the index block is present in an in-memory LRU block cache, fetching a dataset element only needs 1 round trip instead of 2. This is particularly useful when CacheDataset are read from remote file system at a higher latency such as HDFS and GCS.

Almost all code are imported from the LevelDB project, in particular the hash function to shard LRU cache. Currently using Hash32 in core/lib/hash fails the EvictionPolicy test.

I only make 2 modifications to the original cache:

1. Alias leveldb::Slice to tensorflow::StringPiece, which transitively aliases to absl::string_view.
2. Switch to tensorflow::mutex for all mutexes.

Ping @jsimsa to review.

jsimsa commented on tensorflow/core/lib/io/cache_test.cc:
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
this should be using the TensorFlow copyright

byronyi replied:
Changed.

jsimsa commented on tensorflow/core/lib/io/cache.h:
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
this should be using the TensorFlow copyright

byronyi replied:
Changed.

jsimsa commented on tensorflow/core/lib/io/cache.cc:
```diff
@@ -0,0 +1,439 @@
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
```
this should be using the TensorFlow copyright

byronyi replied:
Changed.

Commit:
Add block cache for low level table library

This is part of a patch series aiming to improve the
performance of on-disk dataset.cache() (CacheDatasetV2).

Currently CacheDataset uses core/util/tensor_bundle to cache
dataset elements on disks. It uses sorted string table (SST)
to index dataset elements. Unlike checkpoints which do not
have a great number of tensors, caching a large dataset may
incur a greater number of tensors as well as index blocks.

If the index block is present in an in-memory LRU block cache,
fetching a dataset element only needs 1 round trip instead of 2.
This is particularly useful when CacheDataset are read from
remote file system at a higher latency such as HDFS and GCS.

Almost all code are imported from the LevelDB project,
in particular the hash function to shard LRU cache. Currently
using Hash32 in core/lib/hash fails the EvictionPolicy test.

I only make 2 modifications to the original cache:

1. Alias leveldb::Slice to tensorflow::StringPiece, which
   transitively aliases to absl::string_view.
2. Switch to tensorflow::mutex for all mutexes.

Signed-off-by: Bairen Yi <yibairen.byron@bytedance.com>

Commit:
tensor_bundle: Enable cache for metadata table

The index cache is by default disabled unless one set
the TF_TABLE_INDEX_CACHE_SIZE_IN_MB environment variable.

Signed-off-by: Bairen Yi <yibairen.byron@bytedance.com>

byronyi commented:
Gently ping @gbaned for reviewing. 

It seems we have r2.2 branch cut later this month and I would really like this PR to make it into the next release.

byronyi commented:
@gbaned Hi, is there anything I could do to roll this PR forward?

gbaned commented:
> @gbaned Hi, is there anything I could do to roll this PR forward?

@byronyi Sorry for the slow response, it is processing internally. We will let you know if anything needed. Thank you.

rthadur commented on tensorflow/core/lib/io/cache_test.cc:
```diff
@@ -0,0 +1,238 @@
+#include <string>
+#include <vector>
+
+#include "tensorflow/core/platform/coding.h"
```
`third_party/tensorflow/core/lib/io/cache_test.cc:21:10: error: use of private header from outside its module: 'third_party/tensorflow/core/platform/coding.h'; `

rthadur replied:
@byronyi can you please check this error ?

byronyi replied:
@rthadur Fixed. Mind to kickoff CI again?

Commit:
block-cache: Include public coding header instead

Signed-off-by: Bairen Yi <yibairen.byron@bytedance.com>

byronyi commented:
@gbaned I've fixed the review comment. Mind to kick off CI tests again?

byronyi commented:
@gbaned mind to take a look? r2.2 branch cut draws near.

byronyi commented:
I see this is being merged manually. Thanks!
